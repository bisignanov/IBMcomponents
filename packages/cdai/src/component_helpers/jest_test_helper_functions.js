//
// Copyright IBM Corp. 2020, 2020
//
// This source code is licensed under the Apache-2.0 license found in the
// LICENSE file in the root directory of this source tree.
//
import React from 'react';
import { shallow, mount } from 'enzyme';

export const getJSXForComponent = (
  component,
  defaultProps = {},
  customProps = {}
) => {
  return React.createElement(component, { ...defaultProps, ...customProps });
};

export const mountComponent = (jsx, returnShallow = true) => {
  let component, componentInstance;

  if (returnShallow) {
    component = shallow(jsx);
  } else {
    component = mount(jsx);
  }

  componentInstance = component.instance();

  let unmount = () => {
    unmountComponent(component);
  };

  return {
    component,
    componentInstance,
    unmount,
  };
};

export const unmountComponent = (component) => {
  component && component.length > 0 && component.unmount();
};
