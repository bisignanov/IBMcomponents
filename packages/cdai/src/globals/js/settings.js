const settings = {
  idePrefix: 'ide',
};
module.exports = settings;
