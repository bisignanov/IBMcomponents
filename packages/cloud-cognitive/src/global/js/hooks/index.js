export { useWindowScroll } from './useWindowScroll';
export { useWindowResize } from './useWindowResize';
export { useClickOutside } from './useClickOutside';
export { useCreateComponentFocus } from './useCreateComponentFocus';
export { useCreateComponentStepChange } from './useCreateComponentStepChange';
export { usePreviousValue } from './usePreviousValue';
export { useResetCreateComponent } from './useResetCreateComponent';
export { useValidCreateStepCount } from './useValidCreateStepCount';
