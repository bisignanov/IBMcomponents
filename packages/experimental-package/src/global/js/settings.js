import featureFlags from '../../../../feature-flags/feature-flags';

const expPrefix = 'exp';

export { expPrefix };

export { featureFlags };
