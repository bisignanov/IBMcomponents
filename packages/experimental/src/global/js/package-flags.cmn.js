//
// Copyright IBM Corp. 2020, 2020
//
// This source code is licensed under the Apache-2.0 license found in the
// LICENSE file in the root directory of this source tree.
//

const defaultFlags = {
  // by default only released components are set to true
  component: {
    AboutModal: false,
    ActionBar: false,
    APIKeyDownlaoder: false,
    APIKeyModal: false,
    BreadcrumbWithOverflow: false,
    Card: false,
    ContextHeader: false,
    EmptyState: false,
    ExampleComponent: false,
    ExportModal: false,
    ImportModal: false,
    ModifiedTabs: false,
    Notifications: false,
    PageHeader: false,
    RemoveDeleteModal: false,
    SidePanel: false,
    TagSet: false,
    TearSheet: false,
  },

  // feature level flags
  feature: {
    'a-new-feature': false,
  },
};

const enableAll = (section) => {
  const result = {};
  for (const key in section) {
    result[key] = true;
  }
  return result;
};

const checkEnabled = (flags, overrides = {}) => {
  const result = {};
  for (const key in flags) {
    if (overrides[key]?.enableAll) {
      // enable all for this section
      delete overrides[key].enableAll;
      result[key] = enableAll(flags[key]);
    } else {
      result[key] = { ...flags[key], ...(overrides[key] || {}) };
    }
  }
  return result;
};

const packageFlags = () => {
  let initialized = false;
  let flags;

  // doNotUse parameter is intended for dev flags
  // doNotUse.initAgain causes flags to be refreshed which is handy in tests
  // an override section containing enableAll=true turns on everything in that section - handy for storybook
  return (overrides, doNotUse) => {
    const defaults = { ...defaultFlags };

    if (!initialized || doNotUse?.initAgain === true) {
      initialized = true;

      flags = checkEnabled(defaults, overrides);
    }

    return flags;
  };
};

const getPackageFlags = packageFlags();

module.exports = getPackageFlags;
