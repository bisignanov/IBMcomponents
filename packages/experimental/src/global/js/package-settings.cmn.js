//
// Copyright IBM Corp. 2020, 2020
//
// This source code is licensed under the Apache-2.0 license found in the
// LICENSE file in the root directory of this source tree.
//

const defaultFlags = {
  // by default only released components are set to true
  component: {
    AboutModal: false,
    ActionBar: false,
    APIKeyDownlaoder: false,
    APIKeyModal: false,
    BreadcrumbWithOverflow: false,
    Card: false,
    ContextHeader: false,
    EmptyState: false,
    ExampleComponent: false,
    ExportModal: false,
    ImportModal: false,
    ModifiedTabs: false,
    Notifications: false,
    PageActionItem: false,
    PageHeader: false,
    RemoveDeleteModal: false,
    SidePanel: false,
    TagSet: false,
    Tearsheet: false,
    TearsheetNarrow: false,
    TearsheetShell: false,
    WebTerminal: false,
  },

  // feature level flags
  feature: {
    'a-new-feature': false,
  },
};

const applyOverrides = (target, overrides = {}) => {
  for (const key in overrides) {
    if (key !== 'enableAll' && key !== 'disableAll') {
      // enable all is special handled one level up
      if (typeof target[key] === 'object') {
        if (typeof overrides[key] === 'object') {
          const disableAll = (key === 'component' || key === 'feature') && overrides[key].disableAll === true;
          const enableAll = (key === 'component' || key === 'feature') && overrides[key].enableAll === true;

          if (disableAll || enableAll) {
            // special case loop through all component or feature and enable
            for (const childKey in target[key]) {
              target[key][childKey] = enableAll;
            }
          }

          // process everything else working through branch
          applyOverrides(target[key], overrides[key]);
        } else {
          // reject replacing object with primitive
          console.warn(
            `Attempt to replace setting ${key} with primitive type rejected`
          );
        }
      } else {
        // both primitive types replace
        target[key] = overrides[key];
      }
    }
  }
};

const defaultPrefix = 'exp';

const settings = {
  prefix: defaultPrefix,
  flags: {
    component: { ...defaultFlags.component },
    feature: { ...defaultFlags.feature },
  },
  isComponentEnabled: (component, byDefault = false) => {
    const componentName = component?.name || component;
    const flags = byDefault ? defaultFlags : settings.flags;

    return flags.component[componentName];
  },
  isFeatureEnabled: (feature, byDefault = false) => {
    const flags = byDefault ? defaultFlags : settings.flags;
    return flags.feature[feature];
  },
  overrideSettings: (overrides, reset) => {
    if (reset) {
      settings.prefix = defaultPrefix;
      settings.flags = {
        component: { ...defaultFlags.component },
        feature: { ...defaultFlags.feature },
      };
    }
    applyOverrides(settings, overrides);
  },
  replaceWithCare: (replacement) => {
    console.warn(`

Presumably you know what you are doing as ou have replaced package settings:

------
${JSON.stringify(settings)}
------

with:

++++++
${JSON.stringify(replacement)}
++++++

`);

    for (const key in settings) {
      delete settings[key];
    }

    for (const key in replacement) {
      settings[key] = replacement[key];
    }
  },
};

module.exports = settings;
