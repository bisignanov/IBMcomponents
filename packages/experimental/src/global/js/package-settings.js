//
// Copyright IBM Corp. 2020, 2021
//
// This source code is licensed under the Apache-2.0 license found in the
// LICENSE file in the root directory of this source tree.
//

const defaults = {
  prefix: 'exp',
  // by default only released components are set to true
  component: {
    AboutModal: false,
    ActionBar: false,
    ActionBarItem: false,
    APIKeyDownlaoder: false,
    APIKeyModal: false,
    BreadcrumbWithOverflow: false,
    Card: false,
    ContextHeader: false,
    EmptyState: false,
    ErrorEmptyState: false,
    NoDataEmptyState: false,
    NoTagsEmptyState: false,
    NotFoundEmptyState: false,
    NotificationsEmptyState: false,
    UnauthorizedEmptyState: false,
    ExampleComponent: false,
    ExportModal: false,
    ImportModal: false,
    ModifiedTabs: false,
    Notifications: false,
    PageActionItem: false,
    PageHeader: false,
    RemoveDeleteModal: false,
    SidePanel: false,
    TagSet: false,
    Tearsheet: false,
    TearsheetNarrow: false,
    TearsheetShell: false,
    WebTerminal: false,
  },

  // feature level flags
  feature: {
    'a-new-feature': false,
  },
};

const setAll = (flags, enabled) => {
  for (const key in flags) {
    flags[key] = enabled;
  }
};

const settings = {
  // flags
  prefix: defaults.prefix,
  component: { ...defaults.component },
  feature: { ...defaults.feature },
  // methods
  isComponentEnabled: (component, byDefault = false) => {
    const componentName = component?.name || component;
    const flags = byDefault ? defaults : settings;
    return flags.component[componentName];
  },
  isFeatureEnabled: (feature, byDefault = false) => {
    const flags = byDefault ? defaults : settings;
    return flags.feature[feature];
  },
  setAllComponents: (enabled) => {
    setAll(settings.component, enabled);
  },
  setAllFeatures: (enabled) => {
    setAll(settings.feature, enabled);
  },
};

export default settings;
