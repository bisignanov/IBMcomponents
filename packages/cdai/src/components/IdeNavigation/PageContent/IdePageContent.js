//
// Copyright IBM Corp. 2020, 2020
//
// This source code is licensed under the Apache-2.0 license found in the
// LICENSE file in the root directory of this source tree.
//

import React from 'react';
import PropTypes from 'prop-types';
import { settings } from 'carbon-components';
import { idAttribute } from '../../../component_helpers/IDHelper';
import { idePrefix } from '../../../globals/js/settings';
const { prefix } = settings;

const IdePageContent = ({ bleedClassName, children, className }) => {
  let pageContentWrapper = (
    <div
      {...idAttribute(`IdePageContent-container`)}
      className={`${className} ${idePrefix}-page-content-use-width ${prefix}--grid ${idePrefix}--content-alignment`}
    >
      {children}
    </div>
  );

  return bleedClassName ? (
    <div
      {...idAttribute(`IdePageContent-bleed-container`)}
      className={`${idePrefix}-page-content-use-width ${bleedClassName}`}
    >
      {pageContentWrapper}
    </div>
  ) : (
    pageContentWrapper
  );
};

IdePageContent.defaultProps = {
  className: '',
};

IdePageContent.propTypes = {
  /** bleedClassName. If your content required a bleed effect, provide a class name here (with your desired background colour) to apply this effect */
  bleedClassName: PropTypes.string,
  /** children. Standard prop, required to render content inside this component/take advantage of the provided grid and alignment classes */
  children: PropTypes.node,
  /** className. Append your own css class to the grid if required */
  className: PropTypes.string,
};

export default IdePageContent;
