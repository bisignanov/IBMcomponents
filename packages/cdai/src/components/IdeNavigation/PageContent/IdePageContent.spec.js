//
// Copyright IBM Corp. 2020, 2020
//
// This source code is licensed under the Apache-2.0 license found in the
// LICENSE file in the root directory of this source tree.
//

import IdePageContent from './IdePageContent';
import { withBleedClass, withChildren } from './test_assets/testProps.js';
import * as jth from '../../../component_helpers/jest_test_helper_functions.js';

describe('IdePageContent unit tests', () => {
  let component, unmount;

  const mountTestComponent = (defaultProps = {}, props = {}) => {
    let test = jth.mountComponent(
      jth.getJSXForComponent(IdePageContent, defaultProps, props)
    );
    component = test.component;
    unmount = test.unmount;
  };

  afterEach(() => {
    // if unmount defined, call it to clear up any mounted component
    unmount && unmount();
  });

  it('renders the expected structure with no props provided', () => {
    mountTestComponent();
    expect(component.getElements()).toMatchSnapshot();
  });

  it('renders the expected structure with all provided', () => {
    mountTestComponent({ ...withBleedClass, ...withChildren });
    expect(component.getElements()).toMatchSnapshot();
  });
});
