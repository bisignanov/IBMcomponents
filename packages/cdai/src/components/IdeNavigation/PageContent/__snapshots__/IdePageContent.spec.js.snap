// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IdePageContent unit tests renders the expected structure with all provided 1`] = `
[
  <div
    className="ide-page-content-use-width storybook--ide-page-component-bleed"
    data-ui-id="IdePageContent-bleed-container"
  >
    <div
      className=" ide-page-content-use-width bx--grid ide--content-alignment"
      data-ui-id="IdePageContent-container"
    >
      <div
        className="bx--row"
      >
        <div
          className="bx--col"
        >
          Column 1
        </div>
        <div
          className="bx--col"
        >
          Column 2
        </div>
      </div>
    </div>
  </div>,
]
`;

exports[`IdePageContent unit tests renders the expected structure with no props provided 1`] = `
[
  <div
    className=" ide-page-content-use-width bx--grid ide--content-alignment"
    data-ui-id="IdePageContent-container"
  />,
]
`;
